src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java

```diff
@@ -200,76 +188,47 @@ public boolean isCommentNodesRequired() {
 
     @Override
     public void visitToken(DetailAST ast) {
-        final int lineNo = ast.getLineNo();
-        if (illegalLines.get(lineNo) != null
-                && illegalLines.get(lineNo).equals(ast.getColumnNo())) {
-            log(ast, MSG_KEY);
+        final String commentBeginLine = getLine(ast.getLineNo() - 1);
+        final String beforeComment = commentBeginLine.substring(0, ast.getColumnNo());
+        final String commentContent = ast.getFirstChild().getText();
+        final boolean isBeforeCommentLegal = format.matcher(beforeComment).find();
+
+        if (ast.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+            if (!isBeforeCommentLegal && !isCommentMatchLegalPattern(commentContent)) {
+                log(ast, MSG_KEY);
+            }
         }
-    }
-
-    @Override
-    public void beginTree(DetailAST rootAST) {
-        final Map<Integer, TextBlock> cppComments = getFileContents()
-                .getSingleLineComments();
-        final Map<Integer, List<TextBlock>> cComments = getFileContents()
-                .getBlockComments();
-        final Set<Integer> lines = new HashSet<>();
-        lines.addAll(cppComments.keySet());
-        lines.addAll(cComments.keySet());
-
-        for (Integer lineNo : lines) {
-            final String line = getLines()[lineNo - 1];
-            final String lineBefore;
-            final TextBlock comment;
-            if (cppComments.containsKey(lineNo)) {
-                comment = cppComments.get(lineNo);
-                lineBefore = line.substring(0, comment.getStartColNo());
+        else {
+            final DetailAST commentEnd = ast.getFirstChild().getNextSibling();
+            final String commentEndLine = getLine(commentEnd.getLineNo() - 1);
+            final int commentEndColumnNo = commentEnd.getColumnNo();
+            final String afterComment;
+            if (commentEndLine.charAt(commentEndColumnNo) == '*') {
+                afterComment = commentEndLine.substring(commentEndColumnNo + 2);
             }
             else {
-                final List<TextBlock> commentList = cComments.get(lineNo);
-                comment = commentList.get(commentList.size() - 1);
-                lineBefore = line.substring(0, comment.getStartColNo());
-
-                // do not check comment which doesn't end line
-                if (comment.getText().length == 1
-                        && !CommonUtil.isBlank(line
-                            .substring(comment.getEndColNo() + 1))) {
-                    continue;
-                }
+                afterComment = commentEndLine.substring(commentEndColumnNo + 2 + 1);
```

Comment:
`commentEndColumnNo + 2 + 1` looks weird

This magic additions require some clarification

Reply:
clarification added